Richer Recording representations

h-mayorquin opened the pull request:
While working with the memmap issue I found myself wanting that the Recorders displayed more information on my notebook. Things like the number of samples in a better notation would have been nice. This is compunded for recoders with multiple segments where I needed to use the getter methods to find out what I wanted to known (i.e. `sefl.get_num_samples()`)

Here is a PR that expands the representations to include number of samples and adds a second line with a breakdown of the qualities of each segment. When the recording object is monsegment it displays nothing as it was the case with the old representation.

Two other minor human readability improvements:
1) This shows the numbers with coma separated power of 10**3. That is, it shows 1,000,000 instead of 1000000, 1,000 instead of 1000 and so on and so forth.
2) When the time is higher than minutes, hours or days it shows that number in parenthesis but shows nothing if the number is around seconds

Here are some examples:
```python
from spikeinterface.core.generate import GeneratorRecording

print("Recording with two segments, note the comas in between the numbers and the different durations per segment")
print(GeneratorRecording(durations=[10, 5], sampling_frequency=30_000, num_channels=10))
print("\nRecording with one segment should be as before")
print(GeneratorRecording(durations=[10], sampling_frequency=30_000, num_channels=10))
print("\nRecording with larger durations should show how time is represented when larger than minutes")
print(GeneratorRecording(durations=[70, 3900], sampling_frequency=30_000, num_channels=10))
print("\nTimes from a variety of scales")
print(GeneratorRecording(durations=[0.020, 0.500, 3.0, 150], sampling_frequency=30_000, num_channels=10))
print("\nMore variation of times")
print(GeneratorRecording(durations=[6_235, 90_000, 153_252_15], sampling_frequency=30_000, num_channels=10))
print("\nMearec to show how it looks when file_path is provided")
from spikeinterface.core.datasets import download_dataset
from spikeinterface.extractors.neoextractors.mearec import read_mearec

local_path = download_dataset(remote_path="mearec/mearec_test_10s.h5")
recording, sorting = read_mearec(local_path)
print(recording)
```

Output:
```python
Recording with two segments, note the comas in between the numbers and the different durations per segment
GeneratorRecording: 10 channels - 30.0kHz - 2 segments - 450,000 samples - 15.00s - float32 dtype 
                    17.17 MiB
Segments:
Samples   [300,000, 150,000]
Durations [10.00s, 5.00s]
Memory    [11.44 MiB, 5.72 MiB]

Recording with one segment should be as before
GeneratorRecording: 10 channels - 30.0kHz - 1 segments - 300,000 samples - 10.00s - float32 dtype 
                    11.44 MiB

Recording with larger durations should show how time is represented when larger than minutes
GeneratorRecording: 10 channels - 30.0kHz - 2 segments - 119,100,000 samples 
                    3,970.00s (1.00 hours) - float32 dtype - 4.44 GiB
Segments:
Samples   [2,100,000, 117,000,000]
Durations [70.00s (1.00 minutes), 3,900.00s (1.00 hours)]
Memory    [80.11 MiB, 4.36 GiB]

Times from a variety of scales
GeneratorRecording: 10 channels - 30.0kHz - 4 segments - 4,605,600 samples 
                    153.52s (2.00 minutes) - float32 dtype - 175.69 MiB
Segments:
Samples   [600, 15,000, 90,000, 4,500,000]
Durations [0.02s (20.00 ms), 0.50s (500.00 ms), 3.00s, 150.00s (2.00 minutes)]
Memory    [23.44 KiB, 585.94 KiB, 3.43 MiB, 171.66 MiB]

More variation of times
GeneratorRecording: 10 channels - 30.0kHz - 3 segments - 462,643,500,000 samples 
                    15,421,450.00s (178.00 days) - float32 dtype - 16.83 TiB
Segments:
Samples   [187,050,000, 2,700,000,000, 459,756,450,000]
Durations [6,235.00s (1.00 hours), 90,000.00s (25.00 hours), 15,325,215.00s (177.00 days)]
Memory    [6.97 GiB, 100.58 GiB, 16.73 TiB]

Mearec to show how it looks when file_path is provided
MEArecRecordingExtractor: 32 channels - 32.0kHz - 1 segments - 320,000 samples - 10.00s 
                          float32 dtype - 39.06 MiB
Mearec to show how it looks when file_path is provided
MEArecRecordingExtractor: 32 channels - 32.0kHz - 1 segments - 320,000 samples - 10.00s 
                          float32 dtype - 39.06 MiB
  file_path: [/home/heberto/spikeinterface_datasets/ephy_testing_data/mearec/mearec_test_10s.h5](https://file+.vscode-resource.vscode-cdn.net/home/heberto/spikeinterface_datasets/ephy_testing_data/mearec/mearec_test_10s.h5)
```

samuelgarcia commented:
Cool.
I am OK with this.
  * Maybe 4 additional lines when multi segment is a bit too much.
  * Maybe we could also have a summary() method that force the description independtly of the segment number.
  * Maybe also we should have mechanisim when num_seg > 8 too havething like this
```
Recording with larger durations should show how time is represented when larger than minutes
GeneratorRecording: 10 channels - 30.0kHz - 20 segments - 119,100,000 samples 
                    3,970.00s (1.10 hours) - float32 dtype - 4.44 GiB
Segments:
Samples   [2,100,000, 117,000,000, ..., 2,100,000, 2,100,000]
Durations [70.00s (1.17 minutes), 3,900.00s (1.08 hours), ..., 70.00s (1.17 minutes), 3,900.00s (1.08 hours)]
Memory    [80.11 MiB, 4.36 GiB, ..., 80.11 MiB, 4.36 GiB]
```
  * the "," as separator for big number is a bit confusing in list display because it is alos the list separator.
    I would vote for "_" here which is more python feeling even it is hugly.

h-mayorquin commented:
@samuelgarcia  
Examples of how it looks with underscores.

### Recording with two segments, note the comas in between the numbers and the different durations per segment


`GeneratorRecording(durations=[10, 5], sampling_frequency=30_000, num_channels=10)`
GeneratorRecording: 10 channels - 30.0kHz - 2 segments - 450,000 samples - 15.00s - float32 dtype 
                    17.17 MiB
Segments:
Samples   [300_000, 150_000]
Durations [10.00s, 5.00s]
Memory    [11.44 MiB, 5.72 MiB]


---


### Recording with one segment should be as before


`GeneratorRecording(durations=[10], sampling_frequency=30_000, num_channels=10)`
GeneratorRecording: 10 channels - 30.0kHz - 1 segments - 300,000 samples - 10.00s - float32 dtype 
                    11.44 MiB


---


### Recording with larger durations should show how time is represented when larger than minutes


`GeneratorRecording(durations=[70, 3900], sampling_frequency=30_000, num_channels=10)`
GeneratorRecording: 10 channels - 30.0kHz - 2 segments - 119,100,000 samples 
                    3_970.00s (1.10 hours) - float32 dtype - 4.44 GiB
Segments:
Samples   [2_100_000, 117_000_000]
Durations [70.00s (1.17 minutes), 3_900.00s (1.08 hours)]
Memory    [80.11 MiB, 4.36 GiB]


---


### Times from a variety of scales


`GeneratorRecording(durations=[0.020, 0.500, 3.0, 150], sampling_frequency=30_000, num_channels=10)`
GeneratorRecording: 10 channels - 30.0kHz - 4 segments - 4,605,600 samples 
                    153.52s (2.56 minutes) - float32 dtype - 175.69 MiB
Segments:
Samples   [600, 15_000, 90_000, 4_500_000]
Durations [0.02s (20.00 ms), 0.50s (500.00 ms), 3.00s, 150.00s (2.50 minutes)]
Memory    [23.44 KiB, 585.94 KiB, 3.43 MiB, 171.66 MiB]


---


### More variation of times


`GeneratorRecording(durations=[6_235, 90_000, 153_252_15], sampling_frequency=30_000, num_channels=10)`
GeneratorRecording: 10 channels - 30.0kHz - 3 segments - 462,643,500,000 samples 
                    15_421_450.00s (178.49 days) - float32 dtype - 16.83 TiB
Segments:
Samples   [187_050_000, 2_700_000_000, 459_756_450_000]
Durations [6_235.00s (1.73 hours), 90_000.00s (25.00 hours), 15_325_215.00s (177.38 days)]
Memory    [6.97 GiB, 100.58 GiB, 16.73 TiB]


---


### Mearec to show how it looks when file_path is provided


MEArecRecordingExtractor: 32 channels - 32.0kHz - 1 segments - 320,000 samples - 10.00s 
                          float32 dtype - 39.06 MiB
  file_path: [/home/heberto/spikeinterface_datasets/ephy_testing_data/mearec/mearec_test_10s.h5](https://file+.vscode-resource.vscode-cdn.net/home/heberto/spikeinterface_datasets/ephy_testing_data/mearec/mearec_test_10s.h5)

h-mayorquin commented:
I would prefer sometthing that looks like this @samuelgarcia 

### Recording with two segments, note the comas in between the numbers and the different durations per segment


`GeneratorRecording(durations=[10, 5], sampling_frequency=30_000, num_channels=10)`
GeneratorRecording: 10 channels - 30.0kHz - 2 segments - 450,000 samples - 15.00s - float32 dtype 
                    17.17 MiB
Segments:
Samples   300,000 - 150,000
Durations 10.00s - 5.00s
Memory    11.44 MiB - 5.72 MiB


---


### Recording with one segment should be as before


`GeneratorRecording(durations=[10], sampling_frequency=30_000, num_channels=10)`
GeneratorRecording: 10 channels - 30.0kHz - 1 segments - 300,000 samples - 10.00s - float32 dtype 
                    11.44 MiB


---


### Recording with larger durations should show how time is represented when larger than minutes


`GeneratorRecording(durations=[70, 3900], sampling_frequency=30_000, num_channels=10)`
GeneratorRecording: 10 channels - 30.0kHz - 2 segments - 119,100,000 samples 
                    3,970.00s (1.10 hours) - float32 dtype - 4.44 GiB
Segments:
Samples   2,100,000 - 117,000,000
Durations 70.00s (1.17 minutes) - 3,900.00s (1.08 hours)
Memory    80.11 MiB - 4.36 GiB


---


### Times from a variety of scales


`GeneratorRecording(durations=[0.020, 0.500, 3.0, 150], sampling_frequency=30_000, num_channels=10)`
GeneratorRecording: 10 channels - 30.0kHz - 4 segments - 4,605,600 samples 
                    153.52s (2.56 minutes) - float32 dtype - 175.69 MiB
Segments:
Samples   600 - 15,000 - 90,000 - 4,500,000
Durations 0.02s (20.00 ms) - 0.50s (500.00 ms) - 3.00s - 150.00s (2.50 minutes)
Memory    23.44 KiB - 585.94 KiB - 3.43 MiB - 171.66 MiB


---


### More variation of times


`GeneratorRecording(durations=[6_235, 90_000, 153_252_15], sampling_frequency=30_000, num_channels=10)`
GeneratorRecording: 10 channels - 30.0kHz - 3 segments - 462,643,500,000 samples 
                    15,421,450.00s (178.49 days) - float32 dtype - 16.83 TiB
Segments:
Samples   187,050,000 - 2,700,000,000 - 459,756,450,000
Durations 6,235.00s (1.73 hours) - 90,000.00s (25.00 hours) - 15,325,215.00s (177.38 days)
Memory    6.97 GiB - 100.58 GiB - 16.73 TiB


---


### Recording with way too many segments


`GeneratorRecording(durations=[10, 5, 1000, 8.0, 3.0, 40_000, 7.0, 10_000_000.0], sampling_frequency=30_000, num_channels=10)`
GeneratorRecording: 10 channels - 30.0kHz - 8 segments - 301,230,990,000 samples 
                    10,041,033.00s (116.22 days) - float32 dtype - 10.96 TiB
Segments:
Samples   300,000 - 150,000 - 30,000,000 - 240,000 - 90,000 - 1,200,000,000 - 210,000 - 300,000,000,000
Durations 10.00s - 5.00s - 1,000.00s (16.67 minutes) - 8.00s - 3.00s - 40,000.00s (11.11 hours) - 7.00s - 10,000,000.00s (115.74 days)
Memory    11.44 MiB - 5.72 MiB - 1.12 GiB - 9.16 MiB - 3.43 MiB - 44.70 GiB - 8.01 MiB - 10.91 TiB


---


### Mearec to show how it looks when file_path is provided


MEArecRecordingExtractor: 32 channels - 32.0kHz - 1 segments - 320,000 samples - 10.00s 
                          float32 dtype - 39.06 MiB
  file_path: [/home/heberto/spikeinterface_datasets/ephy_testing_data/mearec/mearec_test_10s.h5](https://file+.vscode-resource.vscode-cdn.net/home/heberto/spikeinterface_datasets/ephy_testing_data/mearec/mearec_test_10s.h5)

h-mayorquin commented:
@samuelgarcia 
> * Maybe 4 additional lines when multi segment is a bit too much.
> Maybe we could also have a summary() method that force the description independtly of the segment number.

Mmm I am slightly against this so maybe @alejoe91 can break the vote.  I can think in two scenarios where users see this information:
* When they run print(recording)
* When they run the recoder in a cell in a notebook by itself (or a the last line of the notebook).

In both of those cases the users are requesting for information about the object. I am usually concerned about displaying information when it is not requested (verbose by default) but here this is what users are asking for. I don't feel that information of every segment is overwhelming and I think is fine to give them that level of detail if they are requesting. For most cases, single segment the information remains the same as it was before.

That said, if you guys want to move in that direction note the following:

- print(object): Calls __str__ method if defined, otherwise falls back to __repr__.
- Executing object in a notebook or IPython shell: Calls __repr__ method.

We could make one them display the shorter / summary information.


> * Maybe also we should have mechanisim when num_seg > 8 too havething like this
I am strongly against this. Just because I don't think it is a common use case and would complicate the code too much.

> * the "," as separator for big number is a bit confusing in list display because it is alos the list separator.
>   I would vote for "_" here which is more python feeling even it is hugly.
I agree with this. I put your suggestion and another suggestion fo mine which I prefer above. H-ugly is a nice portmanteau btw!

alejoe91 commented:
Hi guys, looks great! Two comments:

- I vote for `'` instead of `,`
- If more than 8 segments, then we should print: `first four | ... | last four` for durations and sizes

samuelgarcia commented:
I push a small patch to have this
```python
rec = GeneratorRecording(durations=[0.020, 0.500, 3.0, 150], sampling_frequency=30_000, num_channels=10)
print(rec)
```
```bash
GeneratorRecording: 10 channels - 30.0kHz - 40 segments - 46,056,000 samples 
                    1,535.20s (25.59 minutes) - float32 dtype - 1.72 GiB
Segments:
Samples:   600 | 15,000 | 90,000 ... 15,000 | 90,000 | 4,500,000
Durations: 0.02s (20.00 ms) | 0.50s (500.00 ms) | 3.00s ... 0.50s (500.00 ms) | 3.00s | 150.00s (2.50 minutes)
Memory:    23.44 KiB | 585.94 KiB | 3.43 MiB ... 585.94 KiB | 3.43 MiB | 171.66 MiB
```

h-mayorquin commented:
@samuelgarcia That looks good to me. For the durations it would be great to have a divider after the elipsis (...).

Rigth now we have:
 0.02s (20.00 ms) | 0.50s (500.00 ms) | 3.00s ... 0.50s (500.00 ms) | 3.00s | 150.00s (2.50 minutes)
I think it would be better if it was:
 0.02s (20.00 ms) | 0.50s (500.00 ms) | 3.00s | ... | 0.50s (500.00 ms) | 3.00s | 150.00s (2.50 minutes)

samuelgarcia commented:
yep

h-mayorquin commented:
This looks good to me. I did two further changes after sam editing:
1) I changed a bit the section I did before to improve readability.
2) I changed the functions to get memory to conform to the API:

What do I mean?
Now we have:
* `get_num_samples` and  `get_total_samples` when the first is meant to access the segment properties and the second gives the recording property.
* You also had `get_total_duration` which applies to the whole segment (I added `get_duration` to apply over a single segment).
* Finally, now the memory methods are like that, we have:
`get_memory_size` and `get_total_memory_size` where the first applies to a segment and the second gives the property of the recoder.
